django_ontruck/denormalisation.py

```diff
@@ -0,0 +1,347 @@
+from importlib import import_module
+from re import split
+from uuid import uuid4
+
+from django.db.models import Model, OneToOneField, CASCADE, prefetch_related_objects
+from django.db.models.signals import post_save
+from more_itertools import flatten
+
+
+def capitalize(string):
+    if string == '':
+        return ''
+
+    return '{capital}{rest}'.format(
+        capital=string[0].capitalize(), rest=string[1:]
+    )
+
+
+def camel_case(string):
+    return ''.join([capitalize(x) for x in split(r'[\s_]', string)])
+
+
+class InitialisedOnSetName:
+    def __init__(self):
+        self.attribute_name = '_var_{uuid}'.format(uuid=uuid4())
+
+    def __get__(self, instance, owner):
+        if hasattr(instance, self.attribute_name):
+            return getattr(instance, self.attribute_name)
+
+        raise AttributeError(
+            '{owner} must be used as a descriptor attached to a model_class'.format(
+                owner=owner
+            )
+        )
+
+    def __set__(self, instance, value):
+        setattr(instance, self.attribute_name, value)
+
+
+class DenormalisedProperty(property):
+    def __init__(self, fget, field, prefetches):
+        self.name = fget.__name__
+        self.field = field
+        self.prefetches = prefetches
+
+        super().__init__(fget=fget)
+
+
+class Denormalised:
+    """
+    A descriptor class that is used to control denormalised data.
+
+    :param model_base: The Django model_class base class that the denormalisation
+    model_class will inherit from.
+
+    :param signals: An iterable containing signals which should trigger a
+    re-denormalisation of the model_class.
+
+    ## Example
+
+    class Foo(BaseModel):
+        bar = IntegerField(default=0)
+        baz = IntegerField(default=3)
+
+        denormalised = Denormalised(
+            BaseModel, signals=(
+                post_save,
+                DecoratedSignal(
+                    SomeEvent.signal,
+                    fields=('my_expensive_prop',)
+                )
+            )
+        )
+
+        @denormalised.property(IntegerField())
+        def my_expensive_prop(self):
+            # Some expensive calculation here
+
+    foo = Foo.objects.last()
+
+    # re-denormalise the denormalised_properties of foo
+    Foo.denormalised.handler(None, instance=foo)
+
+    # calculate value of `my_expensive_prop`
+    foo.my_expensive_prop
+
+    # Or use the denormalised value
+    foo.denormalised.my_expensive_prop
+    """
+    name = InitialisedOnSetName()
+    model_class = InitialisedOnSetName()
+    parent_model_class = InitialisedOnSetName()
+
+    def __init__(
+        self,
+        model_base=Model,
+        signals=(post_save,)
+    ):
+        self.model_base = model_base
+        self.denormalised_properties = []
+        self.signals = signals
+
+    @property
+    def related_name(self):
+        return '_denormalised_{}'.format(self.name)
+
+    # Equivalent to __set_name__ in Python 3.6+
```

Comment:
We can use the new features of python 3.8, no need to reimplement it

Reply:
I think this would break our CI, our test matrix includes Python 3.5:

![image](https://user-images.githubusercontent.com/61272684/95179214-a0556b80-07c0-11eb-974b-9780d2c977b1.png)